presto-main/src/main/java/com/facebook/presto/sql/planner/SubfieldUtils.java

```diff
@@ -14,78 +14,84 @@
     {
         return expression instanceof DereferenceExpression || expression instanceof SubscriptExpression;
     }
 
-    public static SubfieldPath subfieldToSubfieldPath(Node expr)
+    public static SubfieldPath deferenceOrSubscriptExpressionToPath(Node expression)
     {
-        ArrayList<SubfieldPath.PathElement> steps = new ArrayList();
+        ImmutableList.Builder<SubfieldPath.PathElement> elements = ImmutableList.builder();
         while (true) {
-            if (expr instanceof SymbolReference) {
-                SymbolReference symbolReference = (SymbolReference) expr;
-                steps.add(new SubfieldPath.PathElement(symbolReference.getName(), 0));
+            if (expression instanceof SymbolReference) {
+                elements.add(new SubfieldPath.NestedField(((SymbolReference) expression).getName()));
                 break;
             }
-            else if (expr instanceof DereferenceExpression) {
-                DereferenceExpression dereference = (DereferenceExpression) expr;
-                steps.add(new SubfieldPath.PathElement(dereference.getField().getValue(), 0));
-                expr = dereference.getBase();
+
+            if (expression instanceof DereferenceExpression) {
+                DereferenceExpression dereference = (DereferenceExpression) expression;
+                elements.add(new SubfieldPath.NestedField(dereference.getField().getValue()));
+                expression = dereference.getBase();
             }
-            else if (expr instanceof SubscriptExpression) {
-                SubscriptExpression subscript = (SubscriptExpression) expr;
+            else if (expression instanceof SubscriptExpression) {
+                SubscriptExpression subscript = (SubscriptExpression) expression;
                 Expression index = subscript.getIndex();
                 if (index instanceof LongLiteral) {
-                    LongLiteral literal = (LongLiteral) index;
-                    steps.add(new SubfieldPath.PathElement(null, literal.getValue(), true));
+                    elements.add(new SubfieldPath.LongSubscript(((LongLiteral) index).getValue()));
                 }
                 else if (index instanceof StringLiteral) {
-                    StringLiteral literal = (StringLiteral) index;
-                    steps.add(new SubfieldPath.PathElement(literal.getValue(), 0, true));
+                    elements.add(new SubfieldPath.StringSubscript(((StringLiteral) index).getValue()));
+                }
+                else if (index instanceof GenericLiteral) {
+                    GenericLiteral literal = (GenericLiteral) index;
+                    if (BIGINT.getTypeSignature().equals(TypeSignature.parseTypeSignature(literal.getType()))) {
+                        elements.add(new SubfieldPath.LongSubscript(Long.valueOf(literal.getValue())));
+                    }
+                    else {
+                        return null;
```

Comment:
Could this cause NPE on any of `deferenceOrSubscriptExpressionToPath`'s callsites?

Reply:
@tdcmeehan It might. Some callers check for null and some don't. This is an existing behavior, hence, I'd like to revisit it in some future PR.